docs/design/configs.md

```diff
@@ -0,0 +1,231 @@
+
+### Configuration
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components with the object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as either environment variables (``EnvVarSource``) or as a new volume (``ConfigDataVolumeSource``).
+5. API server is watched for modifications to the object.
+6. After a modification, actions 3, 4, and 5 are repeated.
+
+**Note**: Environment variables specified in ``ConfigData`` will override the environment variables
+with the same name found in the system if consumed as environment variables.
+
+### Use Case
+
+#### Pod Configuration
+
+```json
+{
+    "apiVersion": "v1",
+	"name": "PodConfiguration2",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+		"DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+``PodConfiguration1`` is intended to be consumed as environment variable key/value pairs.
+
+```json
+{
+    "apiVersion": "v1",
+	"name": "PodConfiguration2",
+    "kind": "ConfigData",
+    "data": {
+        "config.yaml": "CONFIG_YAML_STRING_CONTENTS",
+        "spec.yaml": "SPEC_YAML_STRING_CONTENTS"
+    }
+}
+```
+``PodConfiguration2`` is intended to be used as a mounted volume containing two files with their respective contents.
+
+Then, configuration objects are persisted to the API server as follows:
+```bash
+kubectl create -f pod-configuration-1.json 
+kubectl create -f pod-configuration-2.json
+```
+
+We create a replication controller specification that creates pods using these two configuration objects:
+```yaml
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: ExampleReplicationController
+spec:
+  replicas: 2
+  selector:
+    app: dummy-app
+  template:
+    metadata:
+      labels:
+        app: dummy-app
+    spec:
+      containers:
+      - name: dummy-app
+        image: dummy-app
+        ports:
+        - containerPort: 80
+	# New pods will be created using the ConfigData objects.
+	# Each config object will be updated be it a volume or environment variables.
+	# But, some of them might require a restart on update.
+	# Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
+	#		If both of them contains a same named environment variable, Env2 will be the final one.
+	#		If Env1 is updated, the reload process will be Env1 first, then Env2.
+	#		Even though Env2 is not updated, the 'env' ConfigData objects are reloaded in the
+	#		order they are specified to prevent Env1 from overriding any of the variables in Env2.
+	config-data:
+		env:
+		# PodConfiguration1 is loaded as environment variables into the
+		# main container of the pod. It requires a restart when the object is updated
+	    - name: PodConfiguration1
+		  on_update: restart
```

Comment:
I think it's more appropriate to handle restarting when ConfigData changes as part of the deployments work that's in progress.